jwst/linearity/tests/test_linearity.py

```diff
@@ -0,0 +1,596 @@
+
+
+def test_saturation():
+    """Check that correction is not applied for groups flagged as SATURATED in GROUPDQ"""
+
+    # size of integration
+    nints = 1
+    ngroups = 160
+    xsize = 1032
+    ysize = 1024
+
+    # create a JWST datamodel for MIRI data
+    im = make_rampmodel(nints, ngroups, ysize, xsize)
+    im.meta.instrument.detector = 'MIRIMAGE'
+
+    # set groupdq pixels to saturated
+    im.groupdq[0, 100:, 200, 150] = dqflags.pixel['SATURATED']  # saturated dq flag
+    im.data[0, 150, 200, 150] = 1000.0  # value shouldn't change
+
+    # run through Linearity pipeline
+    outfile = LinearityStep.call(im)
+
+    assert(outfile.data[0, 150, 200, 150] == 1000.0)  # pixel flagged as saturated, shouldn't change
+
+
+def test_nolincorr():
+    """Check that correction is not applied for pixels flagged NO_LIN_CORR in the DQ array
+    of the reference file."""
+
+    # size of integration
+    nints = 1
+    ngroups = 10
+    xsize = 1032
+    ysize = 1024
+
+    # create a JWST datamodel for MIRI data
+    im = make_rampmodel(nints, ngroups, ysize, xsize)
+    im.meta.instrument.detector = 'MIRIMAGE'
+
+    # set data value
+    im.data[0, 5, 500, 500] = 35
+
+    # Create reference file
+    dq = np.zeros((1024, 1032), dtype=int)
+    numcoeffs = 3
+
+    # set reference file DQ to 'NO_LIN_CORR'
+    dq[500, 500] = dqflags.pixel['NO_LIN_CORR']
+
+    ref_model = LinearityModel()
+    ref_model.data = np.zeros(shape=(numcoeffs, 1024, 1032), dtype=np.float32)
+    ref_model.dq = dq
+
+    ref_model.meta.instrument.name = 'MIRI'
+    ref_model.meta.instrument.detector = 'MIRIMAGE'
+    ref_model.meta.subarray.xstart = 1
+    ref_model.meta.subarray.xsize = 1032
+    ref_model.meta.subarray.ystart = 1
+    ref_model.meta.subarray.ysize = 1024
+
+    # run through pipeline (saturation and linearity steps)
+    outfile = lincorr(im, ref_model)
+
+    assert(outfile.pixeldq[500, 500] == dqflags.pixel['NO_LIN_CORR'])
+    assert(outfile.data[0, 5, 500, 500] == 35)  # NO_LIN_CORR, sci value should not change
+
+
+def test_pixeldqprop():
+    """Check that linearity reference file DQ values are populated into the PIXELDQ array of the data
+    This does not fully test mapping, as we have not provided a dq_def to define the
+    mapping of dq flags in the reference file. it's a straightforward mapping (1-1)"""
+
+    # size of integration
+    nints = 1
+    ngroups = 10
+    xsize = 1032
+    ysize = 1024
+
+    # create a JWST datamodel for MIRI data
+    im = make_rampmodel(nints, ngroups, ysize, xsize)
+    im.meta.instrument.detector = 'MIRIMAGE'
+
+    # Create reference file
+    dq = np.zeros((1024, 1032), dtype=int)
+    numcoeffs = 3
+
+    # set PIXELDQ to 'NO_LIN_CORR'
+    dq[500, 500] = dqflags.pixel['NO_LIN_CORR']
+    dq[550, 550] = dqflags.pixel['DO_NOT_USE']
+    dq[560, 550] = dqflags.pixel['HOT']
+    dq[550, 560] = dqflags.pixel['DEAD']
+    dq[500, 300] = 2049  # Hot and DO_NOT_USE
```

Comment:
Instead of having to know the desired values (2048 and 1 in this case), you could potentially just use:
```
dq[500, 300] = np.bitwise_or(dqflags.pixel['HOT'], dqflags.pixel['DO_NOT_USE'])
```
to do the math for you. You've already got numpy imported as "np", so it should be ready to use.

Reply:
I wasn't sure of the syntax for combining flags in the tests, so agreed to change it.

Reply:
updated in f98cc1e